core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -807,8 +801,8 @@ private[spark] class BlockManager(
         // Now that the block is in either the memory or disk store,
         // tell the master about it.
         info.size = size
-        if (tellMaster) {
-          reportBlockStatus(blockId, info, putBlockStatus)
+        if (tellMaster && info.tellMaster) {
```

Comment:
I think that a big clarifying change would be to internally refactor this so that the `tellMaster` parameter is omitted from all calls and so that the `tellMaster = false` case for cached and reassembled broadcast blocks is inferred automatically from the block id. This would clean up a lot of the function signatures and would make the weird "don't tell the master" exception much more readily apparent.